### PR TITLE
2015 07 27/shmnt.r3

### PR DESCRIPTION
**Pull request by hallyn**
This is a merge request only to get comments and feedback.

It implements hot-plug and hot-unplug of 'disk' type devices.  The 'disk' can be either a block device (loop or drive), or a directory to bind-mount.

There are two main issues to be resolved:

1. when lxd starts, it bind-mounts and then make-rshare's /var/lib/lxd/shmounts.  Subdirectories from here are bind-mounted into each container's /.lxdmounts.  This allows us to pass mounts into a container.  The question is what to do about dangling mounts when lxd stops.  If we do not remove the mount, then rm -rf $lxdpath fails with EBUSY, which is annoying.  If we do remove it, then when re-starting lxd, containers which are already running will not be hotpluggable.

The two best solutions I can think of are:

a) (what I'm doing now) Only umount $lxdpath/shmounts at Stop() when no containers are running.  This means there are cases, especially on kill -9 $lxd, when we have to manually unmount lxdmounts.
b) We could start lxd in a separate mount namespace, so that the mount is automatically cleaned up.  This would mean that anything mounted onto the host after lxd has started would not automatically be visible to lxd if / is not rshared.  It would also prevent lxd passing mounts into a container which was started by a previous lxd run.

I do believe (a) (which I'm doing now) is the best solution, but am open to discussion and especially other ideas.

2. There is a problem with a lxc container calling reboot.  When a disk is passed into a running container, there is no lxc.mount.entry registered with the container monitor.  When the container calls reboot, the container will be re-created from the existing config.  That means that any mount which has been hot-plugged or hot-unplugged will not appear (or reappear) after reboot.  Whereas 'lxc stop c1; lxc start c1' will do the right thing.

Again there seem to be two ways to address this, both of which require changes to lxc and perhaps golxc.

a) We could add support for a mode in lxc where reboot causes a full stop and restart of the container, not a reboot shortcut.  This would have the advantage of fully removing the concern of keeping the lxc container and lxd database in sync.  Otherwise, we'll have to worry about that for every place we hotplug (nics, disks, blockdevs, etc)

b) We could add a set_running_config_item call to the lxc command handler.  This currently does not exist, so it would require support in newer lxc and go-lxc.

At the moment this is simply not addressed - therefore if you hotplug a disk into a container, then call reboot from the container, the disk will not reappear.  If you shutdown and start the container, it will reappear.

### CODE REVIEW

**Comment by tych0** [comment]
Some random thoughts:

1. If we're adding this, it seems to me to be a good point to add some kind of killall API endpoint so that it's easier to clean up.

2. I think we should unshare(). We're not using MS_SHARED-ness of / as a feature right now, and I'm not sure where we would in the future. It seems much nicer to have them automatically cleaned up. Of course, this makes it a little complicated to restart. I'll have a think about what to do there.

**Comment by tych0** [comment]
r.e. restarting, I think we can do the following:

1. unshare() when LXD starts the first time
2. if any containers have started, setns() to (one of) the monitor process' mountns (this is a little bit of a clusterfuck because of setns-ing in go)
3. we can do mounts the user requests as usual

stgraber pointed out that without >= vivid, / is mounted MS_SHARED, so before that we won't be able to mount stuff from mounts that were created after the container was started before then, but I think that's probably okay.

**Comment by stgraber** [comment]
pre-vivid we can also just document that users should add rshared to their / mounting options.

**Comment by hallyn** [comment]
Hm, ok, if we're happy with requiring MS_SHARED it's ok with me.  setns()ing to the mntns of one of the running containers should work - funky but should work :)  will have to be done from the early-init of course.  Would there be any objection to merging this as is while I work on that as a new feature?

I may have misread but I see no response about the rebooting-a-container issue.  If there are no other ideas, I'm thinking we'll leave it as is right now leaving a new open issue, and I'll work on an upstream-lxc 'set-running-config' option.

**Comment by tych0** [comment]
On Tue, Jul 28, 2015 at 07:14:01AM -0700, Serge Hallyn wrote:
> Hm, ok, if we're happy with requiring MS_SHARED it's ok with me.  setns()ing to the mntns of one of the running containers should work - funky but should work :)  will have to be done from the early-init of course.  Would there be any objection to merging this as is while I work on that as a new feature?

Not in principle, although I haven't looked at the branch. I'll do
that now.

> I may have misread but I see no response about the rebooting-a-container issue.  If there are no other ideas, I'm thinking we'll leave it as is right now leaving a new open issue, and I'll work on an upstream-lxc 'set-running-config' option.

Yeah, I think this is probably more a question for @stgraber, but I
think since this matches the behavior of reboot everywhere else (i.e.
config changes are not reflected) it seems ok to me to leave it as an
open problem.

Tycho

File: lxd/daemon.go [hunk 1]
```diff
@@ -539,6 +603,14 @@ func (d *Daemon) Stop() error {
 		socket.Close()
 	}
 
+	if n, err := d.numRunningContainers(); err != nil || n == 0 {
```
Comment (tych0) [inline lxd/daemon.go]: I think this is racy, but using `unshare()` will fix it, so I think we can just leave it for now.

File: shared/util.go [hunk 2]
```diff
@@ -412,3 +412,30 @@ func IntInSlice(key int, list []int) bool {
...
+		if rows[3] != path || rows[4] != path {
+			continue
+		}
+		return strings.HasPrefix(rows[6], "shared:")
+	}
+	return false
+}
+
+/* TODO - fill me in */
+func DetectFsType(path string) (string, error) {
```
Comment (tych0) [inline shared/util.go]: We have `shared.GetFilesystem`, can we use that here instead?

**Comment by hallyn** [comment]
(I'll leave a note when this is again ready to merge;  i'm afraid to close it as i don't want to lose the comments)

**Comment by hallyn** [comment]
I think it is ready.  I'll start on a separate branch to handle the lxd-in-its-own-namespace changes.

**Pull request merged**